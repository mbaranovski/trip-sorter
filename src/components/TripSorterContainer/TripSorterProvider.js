import React from "react";
import TripSelection from "../TripSelection/TripSelection";
import TripSorter from "../../utils/TripSorter";
import TripViewer from "../TripViewer/TripViewer";

export default class TripSorterProvider extends React.Component {
  state = {
    tripSorter: null,
    trips: [],
    cities: [],
    currency: "EUR"
  };

  async componentDidMount() {
    const { deals, currency } = await (await fetch("./response.json")).json();
    const tripSorter = new TripSorter(deals);

    // Does not matter if we build cities list from cheapest or shortest path.
    const cities = Object.keys(tripSorter.cheapestPath.graph);
    this.setState({ tripSorter, cities, currency });
  }

  findTrip = (from, to, mode) => () => {
    if (!from || !to) {
      alert("Destination and arrival city must be first selected.");
      return;
    }
    this.setState({ trips: this.state.tripSorter.findTrip(from, to, mode) });
  };

  handleReset = () => {
    this.setState({ trips: [] });
  };

  render() {
    const { trips, tripSorter, cities, currency } = this.state;
    if (!tripSorter) return <div>Loading...</div>;

    return (
      <React.Fragment>
        <TripSelection cities={cities} findTrip={this.findTrip} />
        {trips && trips.length ? (
          <TripViewer
            handleReset={this.handleReset}
            currency={currency}
            trips={trips}
          />
        ) : null}
      </React.Fragment>
    );
  }
}
