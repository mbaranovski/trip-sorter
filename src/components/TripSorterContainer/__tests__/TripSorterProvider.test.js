import React from "react";

import {
  render,
  wait,
  fireEvent,
  renderIntoDocument
} from "react-testing-library";
import "dom-testing-library/extend-expect";
import TripSorterProvider from "../TripSorterProvider";
const fetchMock = require("fetch-mock");
const fs = require("fs");

fetchMock.mock(
  "./response.json",
  fs.readFileSync("src/data/response.json", "utf8")
);

test("Renders TripSorterProvider component and checks for children's inital UI state", async () => {
  const { queryByTestId, queryByText } = render(<TripSorterProvider />);

  // Before loading fetching deals we expect to show Loading...
  expect(queryByText("Loading...")).toBeInTheDOM();

  // Component fetches deals and renders TripSelection Component
  await wait(() => {
    expect(queryByText("Loading...")).not.toBeInTheDOM();

    expect(queryByTestId("arrivalField-select")).toBeInTheDOM();
    expect(queryByTestId("departureField-select")).toBeInTheDOM();

    expect(queryByText("Cheapest")).toBeInTheDOM();
    expect(queryByText("Fastest")).toBeInTheDOM();
    expect(queryByText("Search")).toBeInTheDOM();
  });
});

test("Find and render trip from Stockholm to Amsterdam", async () => {
  const {
    getByTestId,
    getByText,
    queryByText,
    queryByTestId
  } = renderIntoDocument(<TripSorterProvider />);

  await wait(); // Work around to wait for the async componentDidMount to return

  expect(queryByText("Your trip")).not.toBeInTheDOM();

  const departureField = getByTestId("departureField-select");
  departureField.value = "Stockholm";
  fireEvent.change(departureField);

  const arrivalField = getByTestId("arrivalField-select");
  arrivalField.value = "Amsterdam";
  fireEvent.change(arrivalField);

  fireEvent.click(getByText("Fastest"));
  fireEvent.click(getByText("Search"));

  expect(queryByText("Your trip:")).toBeInTheDOM();

  expect(queryByText("Stockholm - Warsaw")).toBeInTheDOM();
  expect(queryByTestId("cost-0")).toBeInTheDOM();
  expect(queryByTestId("normal-cost-0")).toBeInTheDOM();
  expect(queryByText("TSW0500")).toBeInTheDOM();
  expect(queryByTestId("time-0")).toBeInTheDOM();
  expect(queryByTestId("transport-0")).toBeInTheDOM();

  expect(queryByText("Warsaw - Amsterdam")).toBeInTheDOM();
  expect(queryByTestId("cost-1")).toBeInTheDOM();
  expect(queryByTestId("normal-cost-1")).not.toBeInTheDOM();
  expect(queryByText("CWA0415")).toBeInTheDOM();
  expect(queryByTestId("time-1")).toBeInTheDOM();
  expect(queryByTestId("transport-1")).toBeInTheDOM();
});
