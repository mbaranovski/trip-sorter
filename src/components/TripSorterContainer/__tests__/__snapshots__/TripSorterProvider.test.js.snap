// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders TripSorterProvider component and checks for children's inital UI state 1`] = `
<div>
  Loading...
</div>
`;
